collections/_article/migrate-animation-from-godot-4-0-to-4-3.md

```diff
@@ -0,0 +1,317 @@
+</video>
+
+It creates a condition where the appearance of the animations do not match between them. So, as you can see from the results, the 3D model pipeline required that resetting behavior.
+
+However, it was not only beneficial, but also caused issues in some cases. Especially frequently reported were cases using `AnimationNodeStateMachine` with 2D project.
+
+If there is no crossfade in the `AnimationNodeStateMachine` transition, it appears to be just an animation is switched, but internally it is blending, which causes the resetting. This made confusing because it was appeared to be doing the same thing as AnimationPlayer, but it was behaving differently.
+
+Finally, we added a `Deterministic` option in 4.2 that allows the user to set whether the resetting is enabled or not. Also, now that `Deterministic` blending is available in AnimationPlayer, there are more possibilities for projects that use AnimationPlayer.
+
+**For example, this is useful when animating a GUI:**
+
+#### Non-deterministic behavior
+
+<video autoplay loop muted playsinline>
+  <source src="/storage/blog/migrate-animation-from-godot-4-0-to-4-3/non_deterministic.mp4?1" type="video/mp4">
+</video>
+
+#### Deterministic behavior
+
+<video autoplay loop muted playsinline>
+  <source src="/storage/blog/migrate-animation-from-godot-4-0-to-4-3/deterministic.mp4?1" type="video/mp4">
+</video>
+
+Finally, note that if the `Deterministic` option is disabled, the total weight of blending will be normalized, so additive blending will not work expectedly.
+
+### CallbackModeDiscrete
+
+From what I have seen so far, not many people understand the behavior of the update mode `Discrete`.
+
+When keying `Discrete` values such as `int`, `bool` and other non-numerical type, the `UpdateMode` defaults to `Discrete`.
+
+![discrete option](/storage/blog/migrate-animation-from-godot-4-0-to-4-3/option_discrete.webp)
+
+On the other hand, `InterpolateType` has a `Nearest`.
+
+![nearest option](/storage/blog/migrate-animation-from-godot-4-0-to-4-3/option_nearest.webp)
+
+What is the difference between them?
+
+**Here is a diagram of how they work as the animation progresses:**
+
+![discrete vs nearest keyframe](/storage/blog/migrate-animation-from-godot-4-0-to-4-3/discrete_keyframe-01.webp)
+
+As you can see, the frequency with which object values are updated is different.
+
+**Not only the frequency, but also the timing of the value updates is different within a single frame:**
+
+![discrete vs linear process](/storage/blog/migrate-animation-from-godot-4-0-to-4-3/mixer_process-01.webp)
+
+Since `Discrete` assumes that there is no blending, it sets the value of the object during the blending process, but `Continuous` sets the value to the object after the blending process has been completed. This means that if it simply blend tracks with `Discrete` and `Continuous`, the result of the `Continuous` track will always take precedence over the `Discrete` track.
```

Comment:
```suggestion
Since `Discrete` assumes that there is no blending, it sets the value of the object during the blending process, but `Continuous` sets the value to the object after the blending process has been completed. This means that when blending `Discrete` and `Continuous` tracks, the result of the `Continuous` track will always take precedence over the `Discrete` track.
```